File: 2002B2/VB/WinForms-Threading-ProgressBars/ProgressBars/frmThread.vb
'Option Strict Off

Imports System.Drawing
Imports System.Windows.Forms
Imports System.Threading

Imports System.ComponentModel

Public Class frmThread
    Inherits System.Windows.Forms.Form
    Private oThreads(6) As Thread
    Private oThreadStart As ThreadStart
    Private strThreadState(6) As String
    Private Const STARTED As String = "Started"
    Private Const STOPPED As String = "Stopped"
    Private Const FINISHED As String = "Finished"
    
    Public Sub New()
        MyBase.New()
        
        frmThread = Me
        
        'This call is required by the Win Form Designer.
        InitializeComponent()
        
        'TODO: Add any initialization after the InitializeComponent() call
        'Initialize the ProgressBars
        ProgressBar1.Minimum = 1
        ProgressBar1.Maximum = 50
        ProgressBar2.Minimum = 1
        ProgressBar2.Maximum = 50
        ProgressBar3.Minimum = 1
        ProgressBar3.Maximum = 50
        ProgressBar4.Minimum = 1
        ProgressBar4.Maximum = 50
        ProgressBar5.Minimum = 1
        ProgressBar5.Maximum = 50
        
    End Sub
    
    'Form overrides dispose to clean up the component list.
    Public Overrides Sub Dispose()
        MyBase.Dispose()
        components.Dispose()
    End Sub
    
#Region " Windows Form Designer generated code "
    
    'Required by the Windows Form Designer
    Private components As System.ComponentModel.Container
    Private WithEvents Button6 As System.Windows.Forms.Button
    Private WithEvents Button5 As System.Windows.Forms.Button
    Private WithEvents Button4 As System.Windows.Forms.Button
    Private WithEvents Button3 As System.Windows.Forms.Button
    
    
    
    
    Private WithEvents Button2 As System.Windows.Forms.Button
    
    
    
    Private WithEvents Button1 As System.Windows.Forms.Button
    Private WithEvents ProgressBar5 As System.Windows.Forms.ProgressBar
    Private WithEvents ProgressBar4 As System.Windows.Forms.ProgressBar
    Private WithEvents ProgressBar3 As System.Windows.Forms.ProgressBar
    Private WithEvents ProgressBar2 As System.Windows.Forms.ProgressBar
    Private WithEvents ProgressBar1 As System.Windows.Forms.ProgressBar
    
    Dim WithEvents frmThread As System.Windows.Forms.Form
    
    'NOTE: The following procedure is required by the Windows Form Designer
    'It can be modified using the Windows Form Designer.  
    'Do not modify it using the code editor.
    Private Sub InitializeComponent()
        Me.components = New System.ComponentModel.Container()
        Me.Button4 = New System.Windows.Forms.Button()
        Me.Button3 = New System.Windows.Forms.Button()
        Me.Button2 = New System.Windows.Forms.Button()
        Me.ProgressBar5 = New System.Windows.Forms.ProgressBar()
        Me.ProgressBar4 = New System.Windows.Forms.ProgressBar()
        Me.ProgressBar1 = New System.Windows.Forms.ProgressBar()
        Me.ProgressBar3 = New System.Windows.Forms.ProgressBar()
        Me.ProgressBar2 = New System.Windows.Forms.ProgressBar()
        Me.Button1 = New System.Windows.Forms.Button()
        Me.Button6 = New System.Windows.Forms.Button()
        Me.Button5 = New System.Windows.Forms.Button()
        
        '@design Me.TrayHeight = 0
        '@design Me.TrayLargeIcon = False
        '@design Me.TrayAutoArrange = True
        Button4.Location = New System.Drawing.Point(152, 152)
        Button4.Size = New System.Drawing.Size(64, 23)
        Button4.TabIndex = 8
        Button4.Text = "Suspend"
        
        Button3.Location = New System.Drawing.Point(152, 104)
        Button3.Size = New System.Drawing.Size(64, 23)
        Button3.TabIndex = 7
        Button3.Text = "Suspend"
        
        Button2.Location = New System.Drawing.Point(152, 64)
        Button2.Size = New System.Drawing.Size(64, 23)
        Button2.TabIndex = 6
        Button2.Text = "Suspend"
        
        ProgressBar5.Location = New System.Drawing.Point(16, 232)
        ProgressBar5.TabIndex = 4
        ProgressBar5.Size = New System.Drawing.Size(128, 23)
        
        ProgressBar4.Location = New System.Drawing.Point(16, 192)
        ProgressBar4.TabIndex = 3
        ProgressBar4.Size = New System.Drawing.Size(128, 23)
        
        ProgressBar1.Location = New System.Drawing.Point(16, 64)
        ProgressBar1.TabIndex = 0
        ProgressBar1.Size = New System.Drawing.Size(128, 23)
        
        ProgressBar3.Location = New System.Drawing.Point(16, 152)
        ProgressBar3.TabIndex = 2
        ProgressBar3.Size = New System.Drawing.Size(128, 23)
        
        ProgressBar2.Location = New System.Drawing.Point(16, 104)
        ProgressBar2.TabIndex = 1
        ProgressBar2.Size = New System.Drawing.Size(128, 23)
        
        Button1.Location = New System.Drawing.Point(16, 16)
        Button1.Size = New System.Drawing.Size(88, 32)
        Button1.TabIndex = 5
        Button1.Text = "StartThreads"
        
        Button6.Location = New System.Drawing.Point(152, 232)
        Button6.Size = New System.Drawing.Size(64, 23)
        Button6.TabIndex = 10
        Button6.Text = "Suspend"
        
        Button5.Location = New System.Drawing.Point(152, 192)
        Button5.Size = New System.Drawing.Size(64, 23)
        Button5.TabIndex = 9
        Button5.Text = "Suspend"
        Me.Text = "frmThread"
        Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
        Me.ClientSize = New System.Drawing.Size(296, 285)
        
        Me.Controls.Add(Button6)
        Me.Controls.Add(Button5)
        Me.Controls.Add(Button4)
        Me.Controls.Add(Button3)
        Me.Controls.Add(Button2)
        Me.Controls.Add(Button1)
        Me.Controls.Add(ProgressBar5)
        Me.Controls.Add(ProgressBar4)
        Me.Controls.Add(ProgressBar3)
        Me.Controls.Add(ProgressBar2)
        Me.Controls.Add(ProgressBar1)
    End Sub
    
#End Region
    
    Protected Sub Button6_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles Button6.click
        Call ThreadState(5, Me.Button6)
    End Sub

    Protected Sub Button5_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles Button5.click
        Call ThreadState(4, Me.Button5)
    End Sub

    Protected Sub Button4_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles Button4.click
        Call ThreadState(3, Me.Button4)
    End Sub

    Protected Sub Button3_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles Button3.click
        Call ThreadState(2, Me.Button3)
    End Sub

    Protected Sub Button2_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles Button2.click
        Call ThreadState(1, Me.Button2)
    End Sub

    Private Sub ThreadState(ByVal ThreadName As Integer, ByVal TheButton As Windows.Forms.Button)
        Select Case strThreadState(ThreadName)
            Case STARTED
                strThreadState(ThreadName) = STOPPED
                TheButton.Text = "Resume"
                oThreads(ThreadName).Suspend()
            Case STOPPED
                strThreadState(ThreadName) = STARTED
                TheButton.Text = "Suspend"
                oThreads(ThreadName).Resume()
            Case FINISHED
                'Do nothing
        End Select

    End Sub

    Public Sub NewThread()
        Dim oThread As Thread
        Dim oProgressBar As System.Windows.Forms.ProgressBar
        Dim oRandom As Random
        Dim i As Integer

        'Get a reference to myself
        'We will use this later to get the name of the thread        
        oThread = System.Threading.Thread.CurrentThread()

        'Here we go!
        Debug.Write("NewThread is running on thread" & oThread.Name)

        'Create an instance of a Random object
        'We are seeding the random number generator with
        'The thread number + the current second, which will
        'Give each thread a different sequence of random numbers
        oRandom = New Random(CInt(oThread.Name) + Second(Now))

        'Using the thread's name we get a reference to the
        'ProgressBar it is responsible for updating
        Select Case oThread.Name
            Case "1"
                oProgressBar = ProgressBar1
            Case "2"
                oProgressBar = ProgressBar2
            Case "3"
                oProgressBar = ProgressBar3
            Case "4"
                oProgressBar = ProgressBar4
            Case "5"
                oProgressBar = ProgressBar5
        End Select

        For i = 1 To 50

            Try
                SyncLock oThread
                    'Update the ProgressBar
                    oProgressBar.Value = i
                    'Sleep a random number of milliseconds
                    'NextDouble will return the next random
                    'Number in the sequence

                    oThread.Sleep(Convert.ToInt32(oRandom.NextDouble) * 1000)
                End SyncLock

            Catch
                'sync conflict
            End Try
        Next i
        'We're through so let's announce it        
        Debug.Write("Thread" & oThread.Name & " has exited.")

        'Mark the thread as "finished"
        strThreadState(CInt(oThread.Name)) = FINISHED

    End Sub


    Protected Sub Button1_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles Button1.click
        Dim i As Integer

        For i = 1 To 5
            'Create a ThreadStart object, passing the address of NewThread             
            oThreadStart = New ThreadStart(AddressOf NewThread)
            'clear any existing threads (if the start button was clicked more than once)
            oThreads(i) = Nothing
            'Create a Thread object 
            oThreads(i) = New Thread(oThreadStart)
            'Starting the thread invokes the ThreadStart delegate
            oThreads(i).Name = i.ToString
            strThreadState(i) = STARTED

            oThreads(i).Start()

        Next i

        Debug.Write("All threads started")
    End Sub

End Class

File: 2002B2/CS/WinForms-Threading-ProgressBars/ProgressBars/frmThread.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using System.Threading;

namespace FreeThread
{
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public class Form1 : System.Windows.Forms.Form
	{
		private Thread[] oThreads = new Thread[6];
		//private ThreadStart[] oThreadStart;
		private string[] strThreadState = new string[]{"","","","","",""};	
		private const string STARTED = "Started";
		private const string STOPPED = "Stopped";
		private const string FINISHED = "Finished";
		private System.Windows.Forms.Button Button1;
		private System.Windows.Forms.ProgressBar ProgressBar2;
		private System.Windows.Forms.ProgressBar ProgressBar3;
		private System.Windows.Forms.ProgressBar ProgressBar1;
		private System.Windows.Forms.Button Button5;
		private System.Windows.Forms.Button Button6;
		private System.Windows.Forms.Button Button2;
		private System.Windows.Forms.Button Button3;
		private System.Windows.Forms.Button Button4;
		private System.Windows.Forms.ProgressBar ProgressBar4;
		private System.Windows.Forms.ProgressBar ProgressBar5;
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components;

		public Form1()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//

			 //Initialize the ProgressBars
			ProgressBar1.Minimum = 1;
			ProgressBar1.Maximum = 50;
			ProgressBar2.Minimum = 1;
			ProgressBar2.Maximum = 50;
			ProgressBar3.Minimum = 1;
			ProgressBar3.Maximum = 50;
			ProgressBar4.Minimum = 1;
			ProgressBar4.Maximum = 50;
			ProgressBar5.Minimum = 1;
			ProgressBar5.Maximum = 50;
		}

		private void ThreadState(int ThreadName, System.Windows.Forms.Button TheButton)
		{
			
			switch (strThreadState[ThreadName])
			{
				case STARTED:
					strThreadState[ThreadName] = STOPPED;
					TheButton.Text = "Resume";
					oThreads[ThreadName].Suspend();
					break;
				case STOPPED:
					strThreadState[ThreadName] = STARTED;
					TheButton.Text = "Suspend";
					oThreads[ThreadName].Resume();
					break;
				case FINISHED:
					//do nothing			
					break;
			}
		}

		internal void NewThread()
		//internal void NewThread(Thread oThread, System.Windows.Forms.ProgressBar oProgressBar, Random oRandom, int i)
		{					
			Thread oThread;        
			System.Windows.Forms.ProgressBar oProgressBar = new System.Windows.Forms.ProgressBar();
			Random oRandom;
			
			//Get a reference to myself
			//We will use this later to get the name of the thread        
			oThread = System.Threading.Thread.CurrentThread;


			//Create an instance of a Random object
			//We are seeding the random number generator with
			//The thread number + the current second, which will
			//Give each thread a different sequence of random numbers
			oRandom = new Random(System.Convert.ToInt16(oThread.Name) + System.DateTime.Now.Second);

			//Using the thread's name we get a reference to the
			//ProgressBar it is responsible for updating			
			switch (oThread.Name) 
			{
				case "1":
					oProgressBar = ProgressBar1;
					break;
				case "2":
					oProgressBar = ProgressBar2;
					break;
				case "3":
					oProgressBar = ProgressBar3;
					break;
				case "4":
					oProgressBar = ProgressBar4;
					break;
				case "5":
					oProgressBar = ProgressBar5;			
					break;
			}

			
			for (int j = 1;j < 50; j++)
			{
				//Update the ProgressBar
				oProgressBar.Value = j;
			
				//Sleep a random number of milliseconds
				//NextDouble will return the next random
				//Number in the sequence
				System.Threading.Thread.Sleep(Convert.ToInt32(oRandom.NextDouble()) * 1000);
			}

			//Mark the thread as "finished"
			strThreadState[Convert.ToInt16(oThread.Name)] = FINISHED;
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		public override void Dispose()
		{
			base.Dispose();
			if(components != null)
				components.Dispose();
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
			this.components = new System.ComponentModel.Container();
			this.ProgressBar2 = new System.Windows.Forms.ProgressBar();
			this.ProgressBar3 = new System.Windows.Forms.ProgressBar();
			this.ProgressBar1 = new System.Windows.Forms.ProgressBar();
			this.Button5 = new System.Windows.Forms.Button();
			this.Button6 = new System.Windows.Forms.Button();
			this.Button1 = new System.Windows.Forms.Button();
			this.Button2 = new System.Windows.Forms.Button();
			this.Button3 = new System.Windows.Forms.Button();
			this.Button4 = new System.Windows.Forms.Button();
			this.ProgressBar4 = new System.Windows.Forms.ProgressBar();
			this.ProgressBar5 = new System.Windows.Forms.ProgressBar();
			this.ProgressBar2.Location = new System.Drawing.Point(16, 104);
			this.ProgressBar2.Size = new System.Drawing.Size(128, 23);
			this.ProgressBar2.TabIndex = 1;
			this.ProgressBar3.Location = new System.Drawing.Point(16, 152);
			this.ProgressBar3.Size = new System.Drawing.Size(128, 23);
			this.ProgressBar3.TabIndex = 2;
			this.ProgressBar1.Location = new System.Drawing.Point(16, 64);
			this.ProgressBar1.Size = new System.Drawing.Size(128, 23);
			this.ProgressBar1.TabIndex = 0;
			this.Button5.Location = new System.Drawing.Point(152, 192);
			this.Button5.Size = new System.Drawing.Size(64, 23);
			this.Button5.TabIndex = 9;
			this.Button5.Text = "Suspend";
			this.Button5.Click += new System.EventHandler(this.Button5_Click);
			this.Button6.Location = new System.Drawing.Point(152, 232);
			this.Button6.Size = new System.Drawing.Size(64, 23);
			this.Button6.TabIndex = 10;
			this.Button6.Text = "Suspend";
			this.Button6.Click += new System.EventHandler(this.Button6_Click);
			this.Button1.Location = new System.Drawing.Point(16, 16);
			this.Button1.Size = new System.Drawing.Size(88, 32);
			this.Button1.TabIndex = 5;
			this.Button1.Text = "StartThreads";
			this.Button1.Click += new System.EventHandler(this.Button1_Click);
			this.Button2.Location = new System.Drawing.Point(152, 64);
			this.Button2.Size = new System.Drawing.Size(64, 23);
			this.Button2.TabIndex = 6;
			this.Button2.Text = "Suspend";
			this.Button2.Click += new System.EventHandler(this.Button2_Click);
			this.Button3.Location = new System.Drawing.Point(152, 104);
			this.Button3.Size = new System.Drawing.Size(64, 23);
			this.Button3.TabIndex = 7;
			this.Button3.Text = "Suspend";
			this.Button3.Click += new System.EventHandler(this.Button3_Click);
			this.Button4.Location = new System.Drawing.Point(152, 152);
			this.Button4.Size = new System.Drawing.Size(64, 23);
			this.Button4.TabIndex = 8;
			this.Button4.Text = "Suspend";
			this.Button4.Click += new System.EventHandler(this.Button4_Click);
			this.ProgressBar4.Location = new System.Drawing.Point(16, 192);
			this.ProgressBar4.Size = new System.Drawing.Size(128, 23);
			this.ProgressBar4.TabIndex = 3;
			this.ProgressBar5.Location = new System.Drawing.Point(16, 232);
			this.ProgressBar5.Size = new System.Drawing.Size(128, 23);
			this.ProgressBar5.TabIndex = 4;
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(296, 285);
			this.Controls.AddRange(new System.Windows.Forms.Control[] {this.Button4,
																		  this.Button3,
																		  this.ProgressBar1,
																		  this.ProgressBar3,
																		  this.Button2,
																		  this.ProgressBar2,
																		  this.ProgressBar5,
																		  this.ProgressBar4,
																		  this.Button6,
																		  this.Button5,
																		  this.Button1});
			this.Text = "Form1";

		}
		#endregion

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main() 
		{
			Application.Run(new Form1());
		}

		private void Button1_Click(object sender, System.EventArgs e)
		{  
			for (int i = 0;i <= 5; i++)
			{      
				//clear any existing threads (if the start button was clicked more than once)
				oThreads[i] = null;

				//Create a ThreadStart object, passing the address of NewThread             
				oThreads[i] = new Thread(new ThreadStart(NewThread));

				//Starting the thread invokes the ThreadStart delegate
				oThreads[i].Name = i.ToString();
				strThreadState[i] = STARTED;

				oThreads[i].Start();
			}
		}

		private void Button2_Click(object sender, System.EventArgs e)
		{
			ThreadState(1, this.Button2);
		}

		private void Button3_Click(object sender, System.EventArgs e)
		{
			ThreadState(2, this.Button3);
		}

		private void Button4_Click(object sender, System.EventArgs e)
		{
			ThreadState(3, this.Button4);
		}

		private void Button5_Click(object sender, System.EventArgs e)
		{
			ThreadState(4, this.Button5);
		}

		private void Button6_Click(object sender, System.EventArgs e)
		{
			ThreadState(5, this.Button6);
		}
	}
}
